### DISCUSSION TITLE
wait for a change vector

### DISCUSSION THREAD

**Discussion by olibanjoli** [ravendb/ravendb#14569] [Q&A] [answered]
`session.Advanced.Revisions.GetAsync<T>("someChangeVector")`

i guess this method fails when the change vector is not available on the node. is there a way to wait until a change vector is on the node for a specific document?

**Comment by ayende** [comment, marked as answer]
Can you explain your scenario here? 
You generally can't predict what the change vector of a revision would be.

**Reply by olibanjoli** [reply]
Alright. Our cluster has 3 raven master nodes. There is one rest api call which writes a document in ravendb. After the write we have to send an external system some results based on the current document. The call to the external system takes >2 seconds. Something we can not change unfortunately. To keep the response time to our users <30ms, we do the call to the external system asynchronously. So we dispatch a message to our bus which is then consumed by another "worker".
That worker might not be connected to the same raven node, hence loading the document by id could potentially load the document before it gets the update from the node that received the write.

So i'm looking for a way to make sure I do load the document which has received the write already in the worker without slowing down the writing call (rest api call). `session.Advanced.WaitForReplicationAfterSaveChanges()` doesnt help here as I would loose HA.

Alternatively i could generate the payload of the message which is sent to the external system in the rest api call and pass it through the bus, but i would prefer to keep the data in the database.

**Reply by ayende** [reply]
Easiest way to handle that is directly:

* Try read the revision by change vector.
* If it isn't there, put the message back at the end of the queue.
* If there is `> MaxTime`, throw.

Technically speaking, you can get the replication notifications, but that is not how I would recommend doing that. Far too complex.

**Reply by olibanjoli** [reply]
perfect, that actually matches the current implementation.

i assumed reading revision by change vector returns null (`session.Advanced.Revisions.GetAsync<T>("someChangeVector")`) and does not fail. had to test that as it's not clearly stated in the docs.

```
public async Task Execute(SomeMessage message)
{
    var policy = Policy.Handle<DataNotReplicatedException>().WaitAndRetry(5, retryNumber => TimeSpan.FromMilliseconds(retryNumber * 500));

    await policy.Execute(async () =>
    {
        using (var session = _store.OpenAsyncSession())
        {
            var doc = await session.Advanced.Revisions.GetAsync<SomeDocument>(message.DocumentChangeVector);

            if (doc == null)
            {
                throw new DataNotReplicatedException($"doc not yet replicated {message.DocumentChangeVector}");
            }

            // ossom code here
        }
    });
}
```
